File: multipaz-swiftui/src/iosMain/swift/DefaultToHumanReadable.swift
import Foundation

/// Default implementation for converting a [Reason] to a [ReasonHumanReadable] for display
/// in prompt dialogs.
///
/// Handles ``ReasonHumanReadable`` (passes through as-is) and ``PresentmentUnlockReason``
/// with an appropriate subtitle based on whether the expected input is biometric, PIN, or
/// passphrase. All other ``Reason`` values are mapped to a generic unlock prompt.
///
/// - Parameters:
///   - reason: The reason a secure unlock prompt is being requested.
///   - passphraseConstraints: Optional constraints indicating whether passphrase input is
///     expected, and if so, whether it must be numerical.
/// - Returns: A ``ReasonHumanReadable`` containing localized title/subtitle text for display.
@MainActor
public func defaultToHumanReadable(
    reason: Reason,
    passphraseConstraints: PassphraseConstraints?
) async -> ReasonHumanReadable {
    if let humanReadable = reason as? ReasonHumanReadable {
        return humanReadable
    }
    if reason is PresentmentUnlockReason {
        let subtitle: String
        if passphraseConstraints == nil {
            subtitle = NSLocalizedString(
                "key_unlock_present_bio_subtitle",
                value: "Authentication is required",
                comment: "Subtitle for biometric unlock when presenting a document"
            )
        } else if passphraseConstraints!.requireNumerical {
            subtitle = NSLocalizedString(
                "aks_unlock_present_pin_subtitle",
                value: "Enter the PIN associated with the document",
                comment: "Subtitle for PIN unlock when presenting a document"
            )
        } else {
            subtitle = NSLocalizedString(
                "key_unlock_present_passphrase_subtitle",
                value: "Enter the passphrase associated with the document",
                comment: "Subtitle for passphrase unlock when presenting a document"
            )
        }
        return ReasonHumanReadable(
            title: NSLocalizedString(
                "key_unlock_present_title",
                value: "Verify it's you to share the document",
                comment: "Title for unlock prompt when presenting a document"
            ),
            subtitle: subtitle,
            requireConfirmation: false
        )
    }
    return ReasonHumanReadable(
        title: NSLocalizedString(
            "key_unlock_default_title",
            value: "Verify it's you",
            comment: "Title for generic unlock prompt"
        ),
        subtitle: NSLocalizedString(
            "key_unlock_default_subtitle",
            value: "Authentication is required",
            comment: "Subtitle for generic unlock prompt"
        ),
        requireConfirmation: false
    )
}

File: multipaz-compose/src/commonMain/kotlin/org/multipaz/compose/prompt/DefaultToHumanReadable.kt
package org.multipaz.compose.prompt

import org.jetbrains.compose.resources.getString
import org.multipaz.multipaz_compose.generated.resources.Res
import org.multipaz.multipaz_compose.generated.resources.aks_unlock_present_pin_subtitle
import org.multipaz.multipaz_compose.generated.resources.key_unlock_default_subtitle
import org.multipaz.multipaz_compose.generated.resources.key_unlock_default_title
import org.multipaz.multipaz_compose.generated.resources.key_unlock_present_bio_subtitle
import org.multipaz.multipaz_compose.generated.resources.key_unlock_present_passphrase_subtitle
import org.multipaz.multipaz_compose.generated.resources.key_unlock_present_title
import org.multipaz.presentment.PresentmentUnlockReason
import org.multipaz.prompt.ConvertToHumanReadableFn
import org.multipaz.prompt.Reason
import org.multipaz.securearea.PassphraseConstraints

/**
 * Default implementation of [ConvertToHumanReadableFn] for use in [PromptDialogs].
 *
 * Uses translated string resources from the multipaz-compose module. Handles
 * [Reason.HumanReadable] (passes through as-is) and [PresentmentUnlockReason] with
 * appropriate subtitle based on whether the expected input is biometric, PIN, or passphrase.
 * All other [Reason] values are mapped to a generic unlock prompt.
 *
 * @param unlockReason the reason a secure unlock prompt is being requested.
 * @param passphraseConstraints optional constraints that indicate whether passphrase input is
 * expected, and if so, whether it must be numerical.
 * @return a [Reason.HumanReadable] containing localized title/subtitle text for display.
 */
suspend fun defaultToHumanReadable(
    unlockReason: Reason,
    passphraseConstraints: PassphraseConstraints?
): Reason.HumanReadable = when (unlockReason) {
    is Reason.HumanReadable -> unlockReason
    is PresentmentUnlockReason -> {
        val subtitle = when {
            passphraseConstraints == null -> getString(Res.string.key_unlock_present_bio_subtitle)
            passphraseConstraints.requireNumerical -> getString(Res.string.aks_unlock_present_pin_subtitle)
            else -> getString(Res.string.key_unlock_present_passphrase_subtitle)
        }
        Reason.HumanReadable(
            title = getString(Res.string.key_unlock_present_title),
            subtitle = subtitle,
            requireConfirmation = false
        )
    }

    else -> Reason.HumanReadable(
        title = getString(Res.string.key_unlock_default_title),
        subtitle = getString(Res.string.key_unlock_default_subtitle),
        requireConfirmation = false
    )
}
